packages/flutter_tools/lib/src/android/android_sdk.dart

```diff
@@ -437,10 +431,30 @@ class AndroidSdk {
         ).stdout.trim();
         if (javaHomeOutput.isNotEmpty) {
           final String javaHome = javaHomeOutput.split('\n').last.trim();
-          return fileSystem.path.join(javaHome, 'bin', 'java');
+          return javaHome;
         }
       } on Exception { /* ignore */ }
     }
+    return null;
+  }
+
+  /// First try Java bundled with Android Studio, then sniff JAVA_HOME, then fallback to PATH.
+  static String? findJavaBinary({
```

Comment:
```suggestion
  /// The filesystem path of an appropriate `java` executable.
  ///
  /// This comes from [findJavaHome] if that method returns non-null,
  /// and otherwise from searching PATH.
  static String? findJavaBinary({
```